src/goto-instrument/contracts/instrument_spec_assigns.cpp

```diff
@@ -0,0 +1,559 @@
+            typecast_exprt::conditional_cast(size.value(), signed_size_type()),
+            valid_var,
+            lower_bound_var,
+            upper_bound_var,
+            source_location};
+  };
+
+  UNREACHABLE;
+}
+
+void instrument_spec_assignst::create_snapshot(
+  const car_exprt &car,
+  goto_programt &dest) const
+{
+  // adding pragmas to the location to selectively disable checks
+  // where it is sound to do so
+  source_locationt location_no_checks(car.source_location());
+  disable_pointer_checks(location_no_checks);
+
+  goto_programt result;
+
+  result.add(goto_programt::make_decl(car.valid_var(), location_no_checks));
+
+  result.add(goto_programt::make_assignment(
+    car.valid_var(),
+    simplify_expr(
+      and_exprt{car.condition(),
+                not_exprt{null_pointer(car.target_start_address())}},
+      ns),
+    location_no_checks));
+
+  result.add(
+    goto_programt::make_decl(car.lower_bound_var(), location_no_checks));
+
+  result.add(goto_programt::make_assignment(
+    car.lower_bound_var(), car.target_start_address(), location_no_checks));
+
+  result.add(
+    goto_programt::make_decl(car.upper_bound_var(), location_no_checks));
+
+  result.add(goto_programt::make_assignment(
+    car.upper_bound_var(),
+    simplify_expr(
+      plus_exprt{car.target_start_address(), car.target_size()}, ns),
+    location_no_checks));
+
+  add_pragma_disable_assigns_check(result);
+  dest.destructive_append(result);
+}
+
+exprt instrument_spec_assignst::target_validity_expr(
+  const car_exprt &car,
+  bool allow_null_target) const
+{
+  // If requested, we check that when guard condition is true,
+  // the target's `start_address` pointer satisfies w_ok with the expected size
+  // (or is NULL if we allow it explicitly).
+  // This assertion will be falsified whenever `start_address` is invalid or
+  // not of the right size (or is NULL if we dot not allow it expliclitly).
+  auto result =
+    or_exprt{not_exprt{car.condition()},
+             w_ok_exprt{car.target_start_address(), car.target_size()}};
+
+  if(allow_null_target)
+    result.add_to_operands(null_pointer(car.target_start_address()));
+
+  return simplify_expr(result, ns);
+}
+
+void instrument_spec_assignst::target_validity_assertion(
+  const car_exprt &car,
+  bool allow_null_target,
+  goto_programt &dest) const
+{
+  source_locationt location_no_checks(car.source_location());
+  if(location_no_checks.get_function().empty())
+    location_no_checks.set_function(function_id);
+  disable_pointer_checks(location_no_checks);
+
+  goto_programt result;
+
+  auto t = result.add(goto_programt::make_assertion(
+    target_validity_expr(car, allow_null_target), location_no_checks));
+
+  std::string comment = "Check that ";
+  comment += from_expr(ns, "", car.target());
+  comment += " is valid";
+  if(!car.condition().is_true())
+  {
+    comment += " when ";
+    comment += from_expr(ns, "", car.condition());
```

Comment:
Also, these comments are going to be shown to the user in the console output and logs, so as you mentioned below using the same syntax as the source code it could be a plus.